pkg/serverless/trace/inferredspan/span_enrichment.go

```diff
@@ -0,0 +1,118 @@
+		Endpoint:      attributes.Path,
+		HTTPURL:       httpurl,
+		OperationName: "aws.apigateway.rest",
+		RequestID:     requestContext.RequestID,
+		ResourceNames: resource,
+		Stage:         requestContext.Stage,
+	}
+
+	setSynchronicity(&inferredSpan, attributes)
+}
+
+// EnrichInferredSpanWithAPIGatewayHTTPEvent uses the parsed event
+// payload to enrich the current inferred span. It applies a
+// specific set of data to the span expected from a HTTP event.
+func EnrichInferredSpanWithAPIGatewayHTTPEvent(attributes EventKeys, inferredSpan InferredSpan) {
+	log.Debug("Enriching an inferred span for a HTTP API Gateway")
+	requestContext := attributes.RequestContext
+	http := requestContext.HTTP
+	path := requestContext.RawPath
+	resource := fmt.Sprintf("%s %s", http.Method, path)
+	httpurl := fmt.Sprintf("%s%s", requestContext.Domain, path)
+	startTime := calculateStartTime(requestContext.TimeEpoch)
+
+	inferredSpan.Span.Name = "aws.httpapi"
+	inferredSpan.Span.Service = requestContext.Domain
+	inferredSpan.Span.Resource = resource
+	inferredSpan.Span.Type = "http"
+	inferredSpan.Span.Start = startTime
+	inferredSpan.Span.Meta = map[string]string{
+		Endpoint:      path,
+		HTTPURL:       httpurl,
+		HTTPMethod:    http.Method,
+		HTTPProtocol:  http.Protocol,
+		HTTPSourceIP:  http.SourceIP,
+		HTTPUserAgent: http.UserAgent,
+		OperationName: "aws.httpapi",
+		RequestID:     requestContext.RequestID,
+		ResourceNames: resource,
+	}
+
+	setSynchronicity(&inferredSpan, attributes)
+}
+
+// EnrichInferredSpanWithAPIGatewayWebsocketEvent uses the parsed event
+// payload to enrich the current inferred span. It applies a
+// specific set of data to the span expected from a Websocket event.
+func EnrichInferredSpanWithAPIGatewayWebsocketEvent(attributes EventKeys, inferredSpan InferredSpan) {
+	log.Debug("Enriching an inferred span for a Websocket API Gateway")
+	requestContext := attributes.RequestContext
+	endpoint := requestContext.RouteKey
+	httpurl := fmt.Sprintf("%s%s", requestContext.Domain, endpoint)
+	startTime := calculateStartTime(requestContext.RequestTimeEpoch)
+
+	inferredSpan.Span.Name = "aws.apigateway.websocket"
+	inferredSpan.Span.Service = requestContext.Domain
+	inferredSpan.Span.Resource = endpoint
+	inferredSpan.Span.Type = "web"
+	inferredSpan.Span.Start = startTime
+	inferredSpan.Span.Meta = map[string]string{
+		APIID:            requestContext.APIID,
+		APIName:          requestContext.APIID,
+		ConnectionID:     requestContext.ConnectionID,
+		Endpoint:         endpoint,
+		EventType:        requestContext.EventType,
+		HTTPURL:          httpurl,
+		MessageDirection: requestContext.MessageDirection,
+		OperationName:    "aws.apigateway.websocket",
+		RequestID:        requestContext.RequestID,
+		ResourceNames:    endpoint,
+		Stage:            requestContext.Stage,
+	}
+
+	setSynchronicity(&inferredSpan, attributes)
+}
+
+func setSynchronicity(span *InferredSpan, attributes EventKeys) {
```

Comment:
can we try to avoid mutating the `span` here? Maybe by switching to a pure function, returning a `bool`